File: 32. Longest Valid Parentheses/Solution.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        left,right,m = 0,0,0
        for i in s:
            if i == '(':
                left+=1
            else:
                right+=1
            if left == right:
                m = max(m,left*2)
            elif left<right:
                left = 0
                right = 0
        right,left = 0,0
        for i in reversed(s):
            if i == '(':
                left+=1
            else:
                right+=1
            if left == right:
                m = max(m,left*2)
            elif left>right:
                left = 0
                right = 0
        return m

File: 32. Longest Valid Parentheses/Solution.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        int left=0, right=0, m=0;
        for(int i=0; i<s.size(); i++){
            if(s[i]=='('){
                left++;
            }else{
                right++;
            }
            if(left==right){
                m = max(m,left*2);
            }else if(left<right){
                left = 0;
                right = 0;
            }
        }
        right = 0;
        left = 0;
        for(int i=s.size()-1; i>=0; i--){
            if(s[i]=='('){
                left++;
            }else{
                right++;
            }
            if(left==right){
                m = max(m,left*2);
            }else if(left>right){
                left = 0;
                right = 0;
            }
        }
        return m;
    }
};
